### DISCUSSION TITLE
How to install sktime on Raspberry pi?

### DISCUSSION THREAD

**Discussion by bsash** [alan-turing-institute/sktime#4866] [Q&A]
I want to use sktime for time-series classification inference on a Raspberry Pi. I am not able to install it properly with the required dependencies, especially of numpy and numba. When I run `pip install sktime` it installs correctly and also imports properly. But when I try to use `fit` or `predict` it throws an error.

```
# In python interpreter
import numpy as np
from sktime.classification.kernel_based import RocketClassifier
rocket = RocketClassifier(num_kernels=2000)
X_train = np.array([[[1,2,3,4,4,5]], [[1,2,3,4,5,5]], [[1,2,4,5,6,7]]])
y_train = np.array([0, 0, 1])
rocket.fit(X_train, y_train)
```

> UnsupportedParforsError: Failed in nopython mode pipeline (step: nopython mode backend)
The 'parallel' target is not currently supported on 32 bit hardware.
During: lowering "id=0[LoopNest(index_variable = parfor_index.62, range = (0, X_size0.1, 1)), LoopNest(index_variable = parfor_index.63, range = (0, $50binary_multiply.23, 1))]{335: <ir.Block at /home/sashi/.local/lib/python3.9/site-packages/sktime/transformations/panel/rocket/_rocket_numba.py (180)>}Var($parfor_index_tuple_var.69, _rocket_numba.py:180)" at /home/sashi/.local/lib/python3.9/site-packages/sktime/transformations/panel/rocket/_rocket_numba.py (180)

I believe it because the numpy and numba versions are incompatible.
```
numba             0.52.0
numpy             1.21.1
```

I tried `pip install numba==0.53.0` but install fails with `Failed to build llvmlite`.

My platform is:
```
- Current OS
Distributor ID: Raspbian
Description:    Raspbian GNU/Linux 11 (bullseye)
Release:    11
Codename:   bullseye
11.7
- Kernel
6.1.21-v7+
- Architecture
armv7l
- Model
Raspberry Pi 3 Model B Rev 1.2
```

How to install sktime on raspberry pi?

Any help is greatly appreciated!

**Comment by fkiraly** [comment]
Hm, haven't seen this before!
Although in theory, `sktime` core should not be limited by operating system specifics.

Now this is not a full answer, but I want to add my observation to the thread that both `numpy 1.21.1` and `numba 0.52.0` are quite old versions (`numpy 1.21.1` is July 2021, and `numba 0.52.0` is December 2020) current is `numpy 1.25.1` and `numba 0.57.1`.

Maybe what could help:
* can you print out the versions in your python environment?
* most importantly, what is the python version?
* can you try to identify and explain the reason why the versions of those packages are so old?

**Comment by fkiraly** [comment]
Adding the classification and transformer tag (due to `Rocket`). FYI @angus924 re `Rocket`

**Comment by NMertsch** [comment]
(I just found this issue on the Europython Discord and thought I'd have a look. I have no sktime/numba experience and no Raspberry Pi at hand to try it out).

From the error message "The 'parallel' target is not currently supported on 32 bit hardware", I think `@numba.njit(..., parallel=True)` [here](https://github.com/sktime/sktime/blob/ed5ea2dfd66bab5c333d51c947c822ea8f9ef1c1/sktime/transformations/panel/rocket/_rocket_numba.py#L162) is causing the issue (I assume @bsash uses a 32 bit Raspberry Pi).

You could include at "64-bitness" check like documented [here](https://docs.python.org/3/library/platform.html#platform.architecture) and set `parallel=False` otherwise.
Maybe numba has a more elegant way to check if the `parallel` backend is supported. Also, there might be more places where `parallel=True` is set in sktime.

**Reply by fkiraly** [reply]
I see, so your hypothesis is, @NMertsch, that `parallel=True` doesn't work on 32-bit hardware and will always fail?

Currently, `parallel=True` is hardcoded in the decorators for most `numba` based estimators.

But, if indeed this is the cause, we could intercept the `parallel` argument in `sktime.utils.numba.njit` and force it to `False` on 32-bit hardware following the 64-bitness suggested by @NMertsch.

Would you be able to test from an experimental branch, @bsash?

**Comment by pakal** [comment]
The "Pi 3 Model B", like the PI 4, is supposed to support 64bits OS, so you could have a try with raspberry pi os 64bits (but I've personally never tried 64bits version, YMMV)
